smack-extensions/src/test/java/org/jivesoftware/smackx/muc/MultiUserChatTest.java

```diff
@@ -0,0 +1,211 @@
+    public void tearDown() {
+        if (connection != null) {
+            connection.disconnect();
+            connection = null;
+        }
+    }
+
+    @Test
+    public void testInviteDirectly() throws Throwable {
+        EntityBareJid roomJid = JidCreate.entityBareFrom("room@example.com");
+        EntityBareJid userJid = JidCreate.entityBareFrom("user@example.com");
+
+        AtomicBoolean updateRequestSent = new AtomicBoolean();
+        InvokeDirectlyResponder serverSimulator = new InvokeDirectlyResponder() {
+            @Override
+            void verifyRequest(Message updateRequest) {
+                assertEquals(userJid, updateRequest.getTo(), "The provided JID doesn't match the request!");
+
+                GroupChatInvitation groupChatInvitation = (GroupChatInvitation) updateRequest.getExtension(GroupChatInvitation.NAMESPACE);
+                assertNotNull(groupChatInvitation, "Missing GroupChatInvitation extension");
+                assertEquals(roomJid, groupChatInvitation.getRoomAddress());
+
+                updateRequestSent.set(true);
+            }
+        };
+        serverSimulator.start();
+
+        // Create multi user chat
+        MultiUserChat multiUserChat = multiUserChatManager.getMultiUserChat(roomJid);
+
+        // Call tested method
+        multiUserChat.inviteDirectly(userJid);
+
+        // Wait for processing requests
+        serverSimulator.join(RESPONSE_TIMEOUT_IN_MILLIS);
+        serverSimulator.complete();
+
+        // Check if an error occurred within the simulator
+        final Throwable exception = serverSimulator.getException();
+        if (exception != null) {
+            throw exception;
+        }
+
+        assertTrue(updateRequestSent.get(), "Invite directly request not sent");
+    }
+
+    @Test
+    public void shouldReceiveOfflineInvitation() throws XmppStringprepException {
+        EntityBareJid roomJid = JidCreate.entityBareFrom("room@example.com");
+        EntityFullJid inviterJid = JidCreate.entityFullFrom("inviter@example.com/user1");
+        EntityBareJid inviteeJid = JidCreate.entityBareFrom("invitee@example.com");
+
+        GroupChatInvitation groupChatInvitation = new GroupChatInvitation(JidCreate.entityBareFrom("room@example.com"));
+        Message sentMessage = MessageBuilder.buildMessage(UUID.randomUUID().toString())
+                .from(inviterJid)
+                .to(inviteeJid)
+                .addExtension(groupChatInvitation)
+                .build();
+
+        // Prepare listener to receive a group invitation
+        GroupInvitationListener groupInvitationListener = new GroupInvitationListener() {
+            @Override
+            public void verifyInvitation(XMPPConnection conn, MultiUserChat room, EntityJid inviter,
+                                         String reason, String password, Message message, MUCUser.Invite invitation) {
+                // Check all parameters values.
+                assertSame(connection, conn);
```

Comment:
I fear those asserts won't work, as their exceptions will be not be thrown in the thread that invoked the unit test method.

Reply:
Hmmm you're right, it throws by then we have to wait for the latch to time out, and then we can check the stderr in the test report to diagnose the problem. Any suggestions?

```
> Task :smack-extensions:test

MultiUserChatTest > shouldReceiveOfflineInvitationWithAllOptionalAttributes() FAILED
    java.lang.IllegalStateException: Latch timed out before it reached zero
        at org.jivesoftware.smack.test.util.WaitForPacketListener.waitUntilInvocationOrTimeout(WaitForPacketListener.java:49)
        at org.jivesoftware.smackx.muc.MultiUserChatTest.shouldReceiveOfflineInvitationWithAllOptionalAttributes(MultiUserChatTest.java:248)
```

Reply:
Ah, I can have it save off an exception we check and rethrow in the main thread like the `InvokeDirectlyResponder`